src/GitHub.Exports.Reactive/Collections/TrackingCollection.cs

```diff
@@ -0,0 +1,472 @@
+                    if (comparer(list[i], list[idxSmallest]) < 0)
+                        idxSmallest = i;
+                if (idxSmallest == start) continue;
+                Swap(list, start, idxSmallest);
+
+                // if there's a filter, then it's going to trigger events and we don't need to manually trigger them
+                if (filter == null)
+                {
+                    OnPropertyChanged(new PropertyChangedEventArgs("Item[]"));
+                    OnCollectionChanged(new NotifyCollectionChangedEventArgs(NotifyCollectionChangedAction.Move, this[start], start, idxSmallest));
+                }
+            }
+        }
+
+        void UpdateFilter(Func<T, int, IList<T>, bool> newFilter)
+        {
+            if (filter == null && newFilter == null)
+                return; // nothing to do
+
+            // no more filter, add all the hidden items back
+            if (filter != null && newFilter == null)
+            {
+                for (int i = 0; i < original.Count; i++)
+                {
+                    if (GetIndexOf(original[i]) < 0)
+                        InsertItem(i, original[i]);
+                }
+                original = null;
+                filter = null;
+#if !DISABLE_BACKING_DICT
+                liveListMap.Clear();
+#endif
+                return;
+            }
+
+            // there was no filter before, so the Items collection has everything, grab it
+            if (filter == null)
+                original = new List<T>(Items);
+            else
+            {
+                Items.Clear();
+#if !DISABLE_BACKING_DICT
+                liveListMap.Clear();
+#endif
+            }
+            filter = newFilter;
+
+            RecalculateFilter(0, 0, original.Count);
+        }
+
+        void UpdateFilter(TheAction action, int position, int oldPosition)
+        {
+            if (filter == null)
+                return;
+
+            var obj = original[position];
+            var isIncluded = filter(obj, position, this);
+
+            // adding is simple, just check if it should be visible and stuff it at the end
+            if (action == TheAction.Add)
+            {
+                if (isIncluded)
+                    AddIntoLiveList(obj);
+                return;
+            }
+
+            // for all cases except add, we need to know where the touched
+            // element is going to be located in the live list, so grab that
+            var liveListCurrentIndex = GetLiveListPivot(position);
+
+            //if the element was changed but didn't move, then sorting is correct but we 
+            //need to check the filter anyway to determine whether to hide or show the element.
+            //if that changes, then other elements after it might be affected too, so we need to check
+            //all elements from the live list from the element that was changed to the end
+            if (action == TheAction.None)
```

Comment:
Seems like we should use a `switch` statement here for the next three conditionals.